lib/attr.gi

```diff
@@ -37,3 +37,11 @@ InstallGlobalFunction(DisableAttributeValueStoring, function( attr )
     Info(InfoAttributes + InfoWarning, 2, "Disabling value storing for ",NAME_FUNC(attr));
     SET_ATTRIBUTE_STORING( attr, false);
 end);
+
+CHECK_REPEATED_ATTRIBUTE_SET := function(obj, name, val)
+    if not IsBound(obj!.(name)) then
+        Info(InfoAttributes + InfoWarning, 3, "Attribute ", name, " of ", obj, " is marked as assigned, but it has no value");
+    elif obj!.(name) <> val then
+        Info(InfoAttributes + InfoWarning, 3, name, " of ", obj, " already set to ", obj!.(name), ", cannot be changed to ", val);
```

Comment:
I am somewhat wary about printing `obj` as well as the old/new value here. This can lead to pages of output, or trigger errors of its own, or even recursively run into this very same info message.

Which in turn makes me wonder if this should be an info message at all; perhaps it should be a call to `Error` instead (only called if the assertion level is higher than some threshold, and/or controlled by info levels as it is right now). Because thinking about how I'd deal with my code triggering these warning, the first thing I'd do is to edit this function to replace `Info` by `Error`, so that I get a backtrace to find out where the issue actually happened; in contrast, in my experience the printed object or value usually won't be helpful for me.

I guess in the end, it depends a bit on "where do we want to go with this"? Is our goal to stop people from setting non-mutable attributes to new, different values? Then wouldn't an Error (only at high enough level) be the way to go?
This reminds me of my idea of adding a "strict" mode in which all kinds of checks are stricter (see <https://github.com/gap-system/gap/issues/1191>)...


Anyway: please don't get me wrong, I don't want to hold back this PR over some bike shedding, and in the end, I am fine with this PR as it is as well (well, with the performance concern above addressed); I just wonder what the rationale for Info messages vs. Errors is, as naively, I'd expect the Error to be more useful -- but unlike you, I have not yet actually tried this PR, so my naive thoughts here might be totally offbase!

Reply:
I originally used Error in this PR, but there are lots of places where we do try to change attributes to different values. There are places (Like ?SetParent) where we document that trying to assign a different value will be ignored, and it is done in a bunch of places, so we can't just blindly change it. Therefore, it can't really be put at any (reasonable) assertion level, as it is valid and (reasonably) common behaviour to do these reassignments.

However, there are cases where looking at this uncovers bugs (I already reported one, which was fixed), and we could consider moving towards forbidding / reducing how many occur eventually.

I'd also prefer an error, but I'd have to make up a new kind of thing (like a Strict mode as you mentioned), so I thought an Info was better than nothing, and at least means we know in future where we have to "hook" to get more useful information about.

Reply:
BTW, I actually thought that we do document that calling an attribute setter again always has no effect (which of course is contradicted by your leading example). I can't find this in the manual now, though, but I thought that e.g. HPC-GAP relies on this principle in some parts -- but then perhaps this is simply different in HPC-GAP?

Hmmm.. but actually, I cannot reproduce your example at all. I get this:
```
gap> g := Group(());
Group(())
gap> SetTransitivity(g, 4);
gap> Transitivity(g);
4
gap> SetTransitivity(g, 3);
gap> Transitivity(g);
4
```
which is in line with what I thought was happening.

Now I wonder why you see something different?

Reply:
@ChrisJefferson can you explain the discrepancy between the leading example in this PR versus what I observe instead on master?

Reply:
The example at the top was a terrible typo, which went against the whole point :)

There are some setters which document you can reassign (SetParent). Some setters have an explicit check already to see if the value is changed and error if that happen (SetSize). Also properties won't let you set them the other way (but you can assign the same value multiple times):

```
gap> g := Group(());                                                  
Group(())
gap> SetIsAbelian(g, true);
gap> SetIsAbelian(g, true);
gap> SetIsAbelian(g, false);
Error, property is already set the other way
not in any function at *stdin*:8
type 'quit;' to quit to outer loop
```

Reply:
Note that `Parent` is not an attribute, so `SetParent` is not a good example.